diff --git a/package/src/nrc/nrc-netlink.c b/package/src/nrc/nrc-netlink.c
index 9c75a52..fe8cefb 100644
--- a/package/src/nrc/nrc-netlink.c
+++ b/package/src/nrc/nrc-netlink.c
@@ -31,8 +31,8 @@ static struct nrc *nrc_nw;
 
 
 #ifdef CONFIG_SUPPORT_NEW_NETLINK
-static int nrc_nl_pre_doit(const struct genl_ops *ops,
-			   struct sk_buff *skb, struct genl_info *info)
+static int nrc_nl_pre_doit(const struct genl_split_ops *ops,
+		struct sk_buff *skb, struct genl_info *info)
 #else
 static int nrc_nl_pre_doit(struct genl_ops *ops,
 			   struct sk_buff *skb, struct genl_info *info)
@@ -42,8 +42,8 @@ static int nrc_nl_pre_doit(struct genl_ops *ops,
 }
 
 #ifdef CONFIG_SUPPORT_NEW_NETLINK
-static void nrc_nl_post_doit(const struct genl_ops *ops,
-			     struct sk_buff *skb, struct genl_info *info)
+static void nrc_nl_post_doit(const struct genl_split_ops *ops,
+		struct sk_buff *skb, struct genl_info *info)
 #else
 static void nrc_nl_post_doit(struct genl_ops *ops,
 			     struct sk_buff *skb, struct genl_info *info)
