diff --git a/linux/linux.mk b/linux/linux.mk
index 1db5c6046d..e83a0cf88a 100644
--- a/linux/linux.mk
+++ b/linux/linux.mk
@@ -595,6 +595,12 @@ LINUX_PATCH_DEPENDENCIES += $(foreach ext,$(LINUX_EXTENSIONS),\
 LINUX_PRE_PATCH_HOOKS += $(foreach ext,$(LINUX_EXTENSIONS),\
        $(if $(BR2_LINUX_KERNEL_EXT_$(call UPPERCASE,$(ext))),\
                $(call UPPERCASE,$(ext))_PREPARE_KERNEL))
+               
+define LINUX_POST_BUILD_FIXUP
+       make -C $(@D)/tools/objtool
+endef
+
+LINUX_POST_BUILD_HOOKS += LINUX_POST_BUILD_FIXUP
 
 # Checks to give errors that the user can understand
 
